File: source/clibCoreCs/Models/Complex/routeDictionaryModel.cs
using System;

//========================================================================
// This conversion was produced by the Free Edition of
// Instant C# courtesy of Tangible Software Solutions.
// Order the Premium Edition at https://www.tangiblesoftwaresolutions.com
//========================================================================

using Models;
using Models.Context;
using Models.Entity;
using Contensive.Core.Controllers;

namespace Contensive.Core.Models.Complex
{
	public class routeDictionaryModel
	{
		//
		private const string cacheNameRouteDictionary = "routeDictionary";
		//
		//===================================================================================================
		public static Dictionary<string, BaseClasses.CPSiteBaseClass.routeClass> create(coreClass cpCore)
		{
			Dictionary<string, BaseClasses.CPSiteBaseClass.routeClass> result = new Dictionary<string, BaseClasses.CPSiteBaseClass.routeClass>();
			try
			{
				result = getCache(cpCore);
				if (result == null)
				{
					result = new Dictionary<string, BaseClasses.CPSiteBaseClass.routeClass>();
					string physicalFile = "~/" + cpCore.siteProperties.getText("serverpagedefault", "default.aspx");
					List<string> routesAdded = new List<string>();
					List<string> uniqueRouteList = new List<string>();
					//
					// -- admin route
					string adminRoute = genericController.normalizeRoute(cpCore.serverConfig.appConfig.adminRoute);
					if (!string.IsNullOrEmpty(adminRoute))
					{
						result.Add(adminRoute, new BaseClasses.CPSiteBaseClass.routeClass()
						{
							physicalRoute = physicalFile,
							virtualRoute = adminRoute,
							routeType = BaseClasses.CPSiteBaseClass.routeTypeEnum.admin
						});
						uniqueRouteList.Add(adminRoute);
					}
					//
					// -- remote methods
					List<Contensive.Core.Models.Entity.addonModel> remoteMethods = Contensive.Core.Models.Entity.addonModel.createList_RemoteMethods(cpCore, new List<string>());
					foreach (Contensive.Core.Models.Entity.addonModel remoteMethod in remoteMethods)
					{
						string route = genericController.normalizeRoute(remoteMethod.name);
						if (!string.IsNullOrEmpty(route))
						{
							if (uniqueRouteList.Contains(route))
							{
								cpCore.handleException(new ApplicationException("Route [" + route + "] cannot be added because it is a matches the Admin Route or another Remote Method."));
							}
							else
							{
								result.Add(route, new BaseClasses.CPSiteBaseClass.routeClass()
								{
									physicalRoute = physicalFile,
									virtualRoute = route,
									routeType = BaseClasses.CPSiteBaseClass.routeTypeEnum.remoteMethod,
									remoteMethodAddonId = remoteMethod.id
								});
							}
						}
					}
					//
					// -- link forwards
					List<Models.Entity.linkForwardModel> linkForwards = Models.Entity.linkForwardModel.createList(cpCore, "name Is Not null");
					foreach (Models.Entity.linkForwardModel linkForward in linkForwards)
					{
						string route = genericController.normalizeRoute(linkForward.name);
						if (!string.IsNullOrEmpty(route))
						{
							if (uniqueRouteList.Contains(route))
							{
								cpCore.handleException(new ApplicationException("Link Foward Route [" + route + "] cannot be added because it is a matches the Admin Route, a Remote Method or another Link Forward."));
							}
							else
							{
								result.Add(route, new BaseClasses.CPSiteBaseClass.routeClass()
								{
									physicalRoute = physicalFile,
									virtualRoute = route,
									routeType = BaseClasses.CPSiteBaseClass.routeTypeEnum.linkForward,
									linkForwardId = linkForward.id
								});
							}
						}
					}
					//
					// -- link aliases
					List<Models.Entity.linkAliasModel> linkAliasList = Models.Entity.linkAliasModel.createList(cpCore, "name Is Not null");
					foreach (Models.Entity.linkAliasModel linkAlias in linkAliasList)
					{
						string route = genericController.normalizeRoute(linkAlias.name);
						if (!string.IsNullOrEmpty(route))
						{
							if (uniqueRouteList.Contains(route))
							{
								cpCore.handleException(new ApplicationException("Link Alias route [" + route + "] cannot be added because it is a matches the Admin Route, a Remote Method, a Link Forward o another Link Alias."));
							}
							else
							{
								result.Add(route, new BaseClasses.CPSiteBaseClass.routeClass()
								{
									physicalRoute = physicalFile,
									virtualRoute = route,
									routeType = BaseClasses.CPSiteBaseClass.routeTypeEnum.linkAlias,
									linkAliasId = linkAlias.id
								});
							}
						}
					}
					setCache(cpCore, result);
				}
			}
			catch (Exception ex)
			{
				cpCore.handleException(ex);
			}
			return result;
		}
		//
		//====================================================================================================
		public static void setCache(coreClass cpCore, Dictionary<string, BaseClasses.CPSiteBaseClass.routeClass> routeDictionary)
		{
			cpCore.cache.setContent(cacheNameRouteDictionary, routeDictionary);
		}
		//
		//====================================================================================================
		public static Dictionary<string, BaseClasses.CPSiteBaseClass.routeClass> getCache(coreClass cpCore)
		{
			return cpCore.cache.getObject<Dictionary<string, BaseClasses.CPSiteBaseClass.routeClass>>(cacheNameRouteDictionary);
		}
		//
		//====================================================================================================
		public static void invalidateCache(coreClass cpCore)
		{
			cpCore.cache.invalidateContent(cacheNameRouteDictionary);
		}
	}
}

File: source/clibCoreVb/Models/Complex/routeDictionaryModel.vb


Option Explicit On
Option Strict On

Imports Contensive.Core.Models
Imports Contensive.Core.Models.Context
Imports Contensive.Core.Models.Entity
Imports Contensive.Core.Controllers

Namespace Contensive.Core.Models.Complex
    Public Class routeDictionaryModel
        '
        Private Const cacheNameRouteDictionary As String = "routeDictionary"
        '
        '===================================================================================================
        Public Shared Function create(cpCore As coreClass) As Dictionary(Of String, BaseClasses.CPSiteBaseClass.routeClass)
            Dim result As New Dictionary(Of String, BaseClasses.CPSiteBaseClass.routeClass)
            Try
                result = getCache(cpCore)
                If (result Is Nothing) Then
                    result = New Dictionary(Of String, BaseClasses.CPSiteBaseClass.routeClass)
                    Dim physicalFile As String = "~/" & cpCore.siteProperties.getText("serverpagedefault", "default.aspx")
                    Dim routesAdded As New List(Of String)
                    Dim uniqueRouteList As New List(Of String)
                    '
                    ' -- admin route
                    Dim adminRoute As String = genericController.normalizeRoute(cpCore.serverConfig.appConfig.adminRoute)
                    If (Not String.IsNullOrEmpty(adminRoute)) Then
                        result.Add(adminRoute, New BaseClasses.CPSiteBaseClass.routeClass() With {
                                .physicalRoute = physicalFile,
                                .virtualRoute = adminRoute,
                                .routeType = BaseClasses.CPSiteBaseClass.routeTypeEnum.admin
                            })
                        uniqueRouteList.Add(adminRoute)
                    End If
                    '
                    ' -- remote methods
                    Dim remoteMethods As List(Of addonModel) = addonModel.createList_RemoteMethods(cpCore, New List(Of String))
                    For Each remoteMethod As addonModel In remoteMethods
                        Dim route As String = genericController.normalizeRoute(remoteMethod.name)
                        If (Not String.IsNullOrEmpty(route)) Then
                            If (uniqueRouteList.Contains(route)) Then
                                cpCore.handleException(New ApplicationException("Route [" & route & "] cannot be added because it is a matches the Admin Route or another Remote Method."))
                            Else
                                result.Add(route, New BaseClasses.CPSiteBaseClass.routeClass() With {
                                        .physicalRoute = physicalFile,' & "?remoteMethodAddon=" & genericController.EncodeURL(remoteMethod.name),
                                        .virtualRoute = route,
                                        .routeType = BaseClasses.CPSiteBaseClass.routeTypeEnum.remoteMethod,
                                        .remoteMethodAddonId = remoteMethod.id
                                    })
                            End If
                        End If
                    Next
                    '
                    ' -- link forwards
                    Dim linkForwards As List(Of linkForwardModel) = linkForwardModel.createList(cpCore, "name Is Not null")
                    For Each linkForward As linkForwardModel In linkForwards
                        Dim route As String = genericController.normalizeRoute(linkForward.name)
                        If (Not String.IsNullOrEmpty(route)) Then
                            If (uniqueRouteList.Contains(route)) Then
                                cpCore.handleException(New ApplicationException("Link Foward Route [" & route & "] cannot be added because it is a matches the Admin Route, a Remote Method or another Link Forward."))
                            Else
                                result.Add(route, New BaseClasses.CPSiteBaseClass.routeClass() With {
                                        .physicalRoute = physicalFile,' & "?linkForward=" & genericController.EncodeURL(linkForward.name),
                                        .virtualRoute = route,
                                        .routeType = BaseClasses.CPSiteBaseClass.routeTypeEnum.linkForward,
                                        .linkForwardId = linkForward.id
                                    })
                            End If
                        End If
                    Next
                    '
                    ' -- link aliases
                    Dim linkAliasList As List(Of linkAliasModel) = linkAliasModel.createList(cpCore, "name Is Not null")
                    For Each linkAlias As linkAliasModel In linkAliasList
                        Dim route As String = genericController.normalizeRoute(linkAlias.name)
                        If (Not String.IsNullOrEmpty(route)) Then
                            If (uniqueRouteList.Contains(route)) Then
                                cpCore.handleException(New ApplicationException("Link Alias route [" & route & "] cannot be added because it is a matches the Admin Route, a Remote Method, a Link Forward o another Link Alias."))
                            Else
                                result.Add(route, New BaseClasses.CPSiteBaseClass.routeClass() With {
                                        .physicalRoute = physicalFile,' & "?linkAlias=" & genericController.EncodeURL(linkAlias.name),
                                        .virtualRoute = route,
                                        .routeType = BaseClasses.CPSiteBaseClass.routeTypeEnum.linkAlias,
                                        .linkAliasId = linkAlias.id
                                    })
                            End If
                        End If
                    Next
                    Call setCache(cpCore, result)
                End If
            Catch ex As Exception
                cpCore.handleException(ex)
            End Try
            Return result
        End Function
        '
        '====================================================================================================
        Public Shared Sub setCache(cpCore As coreClass, routeDictionary As Dictionary(Of String, BaseClasses.CPSiteBaseClass.routeClass))
            Call cpCore.cache.setContent(cacheNameRouteDictionary, routeDictionary)
        End Sub
        '
        '====================================================================================================
        Public Shared Function getCache(cpCore As coreClass) As Dictionary(Of String, BaseClasses.CPSiteBaseClass.routeClass)
            Return cpCore.cache.getObject(Of Dictionary(Of String, BaseClasses.CPSiteBaseClass.routeClass))(cacheNameRouteDictionary)
        End Function
        '
        '====================================================================================================
        Public Shared Sub invalidateCache(cpCore As coreClass)
            cpCore.cache.invalidateContent(cacheNameRouteDictionary)
        End Sub
    End Class
End Namespace

